editor/editor_data.cpp

```diff
@@ -1141,6 +1141,27 @@ Ref<Texture2D> EditorData::get_script_icon(const Ref<Script> &p_script) {
 		base_scr = base_scr->get_base_script();
 	}
 
+	// No custom icon was found in the inheritance chain, so check the base
+	// class of the script instead.
+	String base_type;
+	p_script->get_language()->get_global_class_name(p_script->get_path(), &base_type);
+
+	// Check if the base type is an extension-defined type.
+	Ref<Texture2D> ext_icon = extension_class_get_icon(base_type);
+	if (ext_icon.is_valid()) {
+		_script_icon_cache[p_script] = ext_icon;
+		return ext_icon;
+	}
+
+	// Look for the base type in the editor theme.
+	// This is only relevant for built-in classes.
+	const Control *gui_base = EditorNode::get_singleton()->get_gui_base();
+	if (gui_base && gui_base->has_theme_icon(base_type, SNAME("EditorIcons"))) {
+		Ref<Texture2D> theme_icon = gui_base->get_theme_icon(base_type, SNAME("EditorIcons"));
+		_script_icon_cache[p_script] = theme_icon;
+		return theme_icon;
+	}
+
 	// If no icon found, cache it as null.
 	_script_icon_cache[p_script] = Ref<Texture>();
```

Comment:
> Before https://github.com/godotengine/godot/pull/75472 the icon could get cached as null and the script wouldn't be re-parsed, but that's not true anymore.

Well, this line still exists.

Reply:
So if we are to go ahead with this, I would add a comment next to the edited line that removes the validity check mentioning that we are being purposefully eager and ignoring invalid cache for the sake of performance. Probably a FIXME comment that should be.

Reply:
The line existed but the null was unused.
Also not sure what you mean by FIXME, this is intended behavior. If the script does not provide icon, there is no reason to parse it again.

Reply:
Script may be changed after we record this into cache. So if we don't check for validity, an editor restart would be required for the icon to be updated. That's why I put the check there.

> The line existed but the null was unused.

Not sure what you mean by that.

Reply:
https://github.com/godotengine/godot/blob/3710f069293f2fde8afe33fea898c4b36fa5e943/editor/editor_data.cpp#L1117
This basically means that the cache was used only when the icon was valid.